src/components/organisms/BannerAdmin/BannerAdmin.tsx

```diff
@@ -1,89 +1,188 @@
-import React, { useState, useCallback, useRef } from "react";
+import React, { useCallback } from "react";
+import { useForm } from "react-hook-form";
 
 import { makeUpdateBanner } from "api/bannerAdmin";
 
 import { AnyVenue } from "types/venues";
+import { Banner } from "types/banner";
+
+import { useShowHide } from "hooks/useShowHide";
+
+import { Button } from "components/atoms/Button";
+import { Checkbox } from "components/atoms/Checkbox";
+import { InputField } from "components/atoms/InputField";
+
+import "./BannerAdmin.scss";
+import classNames from "classnames";
+import { ConfirmationModal } from "components/atoms/ConfirmationModal/ConfirmationModal";
 
 interface BannerAdminProps {
   venueId?: string;
   venue: AnyVenue;
+  onClose?: () => void;
 }
 
 // @debt This component is almost exactly the same as IframeAdmin, we should refactor them both to use the same generic base component
 //   BannerAdmin is the 'canonical example' to follow when we do this
-export const BannerAdmin: React.FC<BannerAdminProps> = ({ venueId, venue }) => {
-  const existingBannerMessage = venue?.bannerMessage ?? "";
-
-  const textareaFieldRef = useRef<HTMLTextAreaElement>(null);
-  const [error, setError] = useState<string | null>();
+export const BannerAdmin: React.FC<BannerAdminProps> = ({
+  venueId,
+  venue,
+  onClose,
+}) => {
+  const { register, handleSubmit, errors, reset, watch } = useForm<Banner>({
+    mode: "onChange",
+    reValidateMode: "onChange",
+  });
+  const isUrlButtonActive = watch(
+    "isActionButton",
+    venue?.banner?.isActionButton
+  );
 
-  const handleInputChange = useCallback(() => {
-    setError(null);
-  }, []);
+  const {
+    isShown: isShowBannerChangeModal,
+    show: showBannerChangeModal,
+    hide: hideBannerChangeModal,
+  } = useShowHide();
 
   const updateBannerInFirestore = useCallback(
-    (msg?: string) => {
+    (banner?: Banner) => {
       if (!venueId) return;
 
-      makeUpdateBanner(venueId, (errorMsg) => setError(errorMsg))(msg);
+      makeUpdateBanner({ venueId, banner });
+      onClose && onClose();
     },
-    [venueId]
+    [venueId, onClose]
   );
 
-  const saveBanner = useCallback(
-    (e: React.MouseEvent<HTMLButtonElement>) => {
-      e.preventDefault();
+  const saveBanner = async (data: Banner) => {
```

Comment:
Why did you remove `useCallback`?

You should probably use `useAsyncFn` and use it's `loading` value to disable the rest of the form and avoid multiple clicks for submit.

Reply:
Looks like `updateBannerInFirestore` is wrapped in `useCallback` you can remove this function and just rename `updateBannerInFirestore` to `saveBanner`

Reply:
fixed 2431677
tho had to remove usecallback in favor of useasyncfn